server/app/controllers/obs.py

```diff
@@ -102,3 +120,84 @@ def human_readable_obs(obs: Observation) -> Observation:
     obs.tca = obs["tca"].strftime("%Y-%m-%d %H:%M:%S") + ", " + strfdelta(aos_tca_duration, fmt="{M:02}m {S:02}s since AOS")
     obs.los = obs["los"].strftime("%Y-%m-%d %H:%M:%S") + ", " + strfdelta(aos_los_duration, fmt="{M:02}m {S:02}s since AOS")
     return obs
+
+@app.route('/obs/delete/<obs_id>', methods=["GET", "POST"])
+def obs_delete(obs_id: ObservationId = None):
+
+    # First check if such an observation even exists.
+    repository = Repository()
+    observation = repository.read_observation(obs_id)
+    if observation is None:
+        return render_template('obs_delete.html', status="There is no observation %s" % obs_id, obs_id=obs_id)
+
+    # Second, check if the guy is logged in.
+    if not current_user.is_authenticated:
+        return render_template('obs_delete.html', status = "You are not logged in, you can't delete anything.", obs_id=obs_id)
+
+    # Ok, at least this guy is logged in. Let's check who he is.
+    user_id = current_user.get_id()
+
+    # Check if the current user is the owner of the station.
+    station = repository.read_station(observation["station_id"])
+    station_id = station['station_id']
+
+    owners = repository.station_owners(station_id)
+    owner = False
+    for o in owners:
+        if o['id'] == user_id:
+            owner = True
+            break
+
+    if not owner:
+        return render_template('obs_delete.html', status = "You are not the owner of station %s, you can't delete observation %s." % (station.name, obs_id), obs_id=obs_id)
+
+    # If you got that far, this means the guy is logged in, he's the owner and is deleting his own observation.
+
+    status = obs_delete_real(repository, obs_id)
+    return render_template('obs_delete.html', status = status, obs_id=obs_id)
+
+def obs_delete_real(repository: Repository, obs_id: ObservationId):
+
+    # Step 1: delete products
+    files = repository.read_observation_files(obs_id)
+
+    root = app.config["storage"]['image_root']
+
+    status = ""
+
+    for f in files:
```

Comment:
This function contains a lot of repeated code.
My proposal: first create a list with the paths to delete and status callbacks.

```python
items = [[
        # Path components, success callback, error callback
        # Main image
        (
            [f['filename']],
            lambda path: "Product %s deleted successfully." % path,
            lambda path: " Product %s was not found." % path
        ),
        # Thumbnail
        (
            ["thumbs", f['filename']],
            lambda path: "Thumbnail file %s deleted successfully." % path,
            lambda path: "Thumbnail file %s was not found." % path
        ),
        # Charts
        (
            ["charts", "by_time-%s.png" % obs_id],
            lambda path: "Timing charts %s deleted successfully." % path,
            None
        )
    ]
    for f in files]

statuses = []
for components, success, fail in items:
    path = os.path.join(root, *components)
    try:
         os.remove(path)
         status.append(success(path))
    except Exception as ex:
         if fail is not None:
             statuses.append(fail(path))
             statuses.append(repr(ex))

status = " ".join(statuses)
```
         
```

Reply:
This is against the kiss principle. Tried this code and got this:
```
#### items=
#### compoenntes= 
(['ab802ca1-419f-418a-aeea-d99bb9c702aa-0-1c60fdf5-3f18-409d-9834-17f014c608c1_product.png'], <function obs_delete_db_and_disk.<locals>.<listcomp>.<lambda> at 0x7f9f72588310>, <function obs_delete_db_and_disk.<locals>.<listcomp>.<lambda> at 0x7f9f725883a0>)
127.0.0.1 - - [08/Jun/2021 23:09:58] "POST /obs/delete/1276 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/usr/lib/python3.8/posixpath.py", line 82, in join
    for b in map(os.fspath, p):
TypeError: expected str, bytes or os.PathLike object, not list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 2463, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 2449, in wsgi_app
    response = self.handle_exception(e)
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 1866, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/thomson/devel/svarog/server/venv/lib/python3.8/site-packages/Flask-1.1.1-py3.8.egg/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/thomson/devel/svarog/server/app/controllers/obs.py", line 148, in obs_delete
    status = obs_delete_db_and_disk(repository, obs_id)
  File "/home/thomson/devel/svarog/server/app/controllers/obs.py", line 187, in obs_delete_db_and_disk
    path = os.path.join(root, *components)
  File "/usr/lib/python3.8/posixpath.py", line 90, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/lib/python3.8/genericpath.py", line 152, in _check_arg_types
    raise TypeError(f'{funcname}() argument must be str, bytes, or '
TypeError: join() argument must be str, bytes, or os.PathLike object, not 'list'
127.0.0.1 - - [08/Jun/2021 23:09:58] "GET /obs/delete/1276?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 -
127.0.0.1 - - [08/Jun/2021 23:09:58] "GET /obs/delete/1276?__debugger__=yes&cmd=resource&f=jquery.js HTTP/1.1" 200 -
127.0.0.1 - - [08/Jun/2021 23:09:58] "GET /obs/delete/1276?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 200 -
127.0.0.1 - - [08/Jun/2021 23:09:58] "GET /obs/delete/1276?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
```

This code has several problems:
- it's overly complicated for no good reason.
- there will soon be observations with multiple products. However, each observation has exactly 1 thumbnail and exactly 2 pass charts. This lambda approach seems to try to do too many things at once.

I've pushed a code that's roughly based on your concept. It generates `files` list which contain pairs of data `[filename, file type]` and there's only one code that tries to delete it and handles exceptions. Would that be ok? I've tested and it seems to work properly on my test setup.